### PR TITLE
Display any votes where MP differs from party on profile page

### PR DESCRIPTION
**Pull request by struan**
This imports the policy vote from PublicWhip and then uses all the votes of current party members to calculate the party's position on a policy. It then compares that with an MP's position and if it is in opposition then it displays that policy on the MP's profile page.

Connects to #910

### CODE REVIEW

File: classes/Party.php [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+        }
+
+        return $scores;
+    }
+
+    public static function getParties() {
+        $db = new \ParlDB;
+
+        $party_list = $db->query(
+            "SELECT DISTINCT party FROM member"
```
Comment (dracos) [inline classes/Party.php]: Ignore empty party?

File: scripts/generate_party_positions.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+include_once dirname(__FILE__) . '/../www/includes/easyparliament/init.php';
+
+$parties = MySociety\TheyWorkForYou\Party::getParties();
+$policies = new MySociety\TheyWorkForYou\Policies;
+
+$party_count = 0;
+foreach ( $parties as $party ) {
+    if ( !$party ) {
+        continue;
```
Comment (dracos) [inline scripts/generate_party_positions.php]: I think this check should be in getParties. However, I think this check is needed for: Independents, potentially Crossbench, and I guess it needs to treat Labour+Labour/Co-operative as one.

File: classes/Party.php [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+                    ':party' => $this->name,
+                    ':division_id' => $division_id
+                )
+            );
+
+            $num_votes = 0;
+            for ( $j = 0; $j < $votes->rows(); $j++ ) {
+                $vote_dir = $votes->field($j, 'vote');
+                if ( $vote_dir == 'tellno' ) $vote_dir = 'no';
+                if ( $vote_dir == 'tellaye' ) $vote_dir = 'aye';
```
Comment (dracos) [inline classes/Party.php]: Are we including teller votes in someone's summary of their position? Presumably we are (and always have?), just wondering about votes such as http://www.publicwhip.org.uk/division.php?date=2015-07-07&number=41 ...

File: classes/Party.php [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+            $votes = $this->db->query(
+                "SELECT count(*) as num_votes, vote
+                FROM persondivisionvotes
+                JOIN member ON persondivisionvotes.person_id = member.person_id
+                WHERE
+                    party = :party
+                    AND member.house = 1
+                    AND division_id = :division_id
+                    AND left_house = '9999-12-31'
```
Comment (dracos) [inline classes/Party.php]: Just wondering if this fix to only current membership causes any issues.

File: classes/Party.php [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+            $total_votes += $num_votes;
+            $max_score += $num_votes * max( array_values( $weights ) );
+        }
+
+        if ( $total_votes == 0 ) {
+            return null;
+        }
+
+        if ( $max_score == 0 ) {
```
Comment (dracos) [inline classes/Party.php]: If this were true, I guess score would have to be 0 too, then weight would be 1, which doesn't sound right? Is this always caught by the above total_votes == 0?

File: www/includes/easyparliament/templates/html/mp/profile.php [hunk 6]
```diff
@@ -55,11 +55,39 @@
...
+                        </p>
+
+                        <p>
+                        However, <?= $full_name ?> sometimes <b>differs</b> from their party collegues, such as:
+                        </p>
+
+                        <ul class="vote-descriptions">
+                          <?php foreach ($sorted_diffs as $policy_id => $score): ?>
+                            <li>
+                                <?= ucfirst(strip_tags($policies[$policy_id])) ?>. <?= $full_name ?> voted <b><?= $positions[$policy_id]['voted'] ?></b>. Most <?= $party ?> MPs voted <b><?= $party_positions[$policy_id]['position'] ?></b>.
```
Comment (dracos) [inline www/includes/easyparliament/templates/html/mp/profile.php]: Just to note #913 changes contents of positions and will affect this.

**Comment by dracos** [comment]
Some questions, but not really on the code which is fine :)

**Comment by struan** [comment]
`$motion->direction` is the result of the vote ( Majority indicates the vote went with the policy and minority against ) and `$motion->motion->policy_vote` is which direction an MP would vote if they agreed with the policy.

File: classes/Party.php (hunk 3 above)
Comment (struan) [inline classes/Party.php]: Yeah, teller votes are included which will give some odd results but I figure we have to do the calculations the same for it to be a valid comparison.

File: classes/Party.php (hunk 4 above)
Comment (struan) [inline classes/Party.php]: It does a little in that you're not comparing someone with how other members of their party voted at the time but I've taken the view that it's more useful to say how the MP compares to their current colleagues.

My hypothetical justification for this is if at an election the new intake causes a party to shift position on something and your MP doesn't agree with that shift then that's more useful to know than the fact that they voted in line with the previous position of the party.

Comment (dracos) [inline classes/Party.php]: Sounds good, as long as it's explained somewhere.

File: www/docs/mp/index.php [hunk 7]
```diff
@@ -464,6 +464,22 @@
         // Generate limited voting record list
         $data['policyPositions'] = new MySociety\TheyWorkForYou\PolicyPositions($policies, $MEMBER, 6);
 
+        // generate party policy diffs
+        $party = new MySociety\TheyWorkForYou\Party($MEMBER->party());
+        $positions = new MySociety\TheyWorkForYou\PolicyPositions( $policiesList, $MEMBER );
+        if ( $party ) {
```
Comment (dracos) [inline www/docs/mp/index.php]: Is this ever false?

**Comment by dracos** [comment]
One tiny query, otherwise :+1:

**Pull request merged**